### PR TITLE
[FIX] purchase: set product control policy based on purchase bill policy

### PR DESCRIPTION
**Pull request by AH-Yussef**
### Steps to reproduce:
- Install **Purchase** and **Inventory** apps
- Go to **General Settings** > **Purchase** > **Invoicing** > **Bill Control**
- Choose **Ordered quantities** and **Save**
- Go to **Purchase** app > **Products** > **Products**
- Create a **New** product
- Set the product type to **Storable product**
- Go to **Purchase** tab > **Control Policy**
- Notice how the control policy is **On received quantities** although we set it up to be O**n Ordered quantities**

### Discussion:
- I added a compute method to manage the default **purchase method** value based on the res.config.setting **Bill Control**.
- I kept the behavior of setting the `purchase_method` to **receive** when the product type is **service** based on [task-2897867](https://www.odoo.com/web#id=2897867&cids=1&menu_id=4720&action=4043&model=project.task&view_type=form) which is associated with the commit **479a38466751e7bc31b26e5635b9735b37dde81f** and [PR](https://github.com/odoo/odoo/pull/96548)

opw-3552786

### CODE REVIEW

**Comment by Whenrow** [comment]
robodoo r+

File: addons/purchase/tests/test_purchase.py [hunk 1]
```diff
@@ -336,3 +336,27 @@ def test_purchase_not_creating_useless_product_vendor(self):
...
+        self.env['ir.default'].set('product.template', 'purchase_method', 'purchase', company_id=True)
+
+        product = self.env['product.product'].create({'name': 'product_test'})
+        self.assertEqual(product.purchase_method, 'purchase')
+
+        product.update({'detailed_type': 'service'})
+        self.assertEqual(product.purchase_method, 'purchase')
+
+        product.update({'detailed_type': 'product'})
+        self.assertEqual(product.purchase_method, 'purchase')
```
Comment (ticodoo) [inline addons/purchase/tests/test_purchase.py:362]: Hey @AH-Yussef it looks like you made a little mistake here. `'detailed_type': 'product'` is added in stock so you need to put these tests in `purchase_stock`.

Also, there's no reason to use `.update()` in these tests, use `write` instead.

File: addons/purchase/models/product.py [hunk 2]
```diff
@@ -19,11 +19,21 @@ class ProductTemplate(models.Model):
     purchase_method = fields.Selection([
         ('purchase', 'On ordered quantities'),
         ('receive', 'On received quantities'),
-    ], string="Control Policy", help="On ordered quantities: Control bills based on ordered quantities.\n"
-        "On received quantities: Control bills based on received quantities.", default="receive")
+    ], string="Control Policy", compute='_compute_purchase_method', default='receive', store=True, readonly=False,
```
Comment (vava-odoo) [inline addons/purchase/models/product.py:22]: @Whenrow @ticodoo 
Shouldn't the default be a lambda here, with a default method with the same behaviour as the compute?
I guess we'd like service products to have a 'purchase' method by default.

Comment (ticodoo) [reply addons/purchase/models/product.py:22]: Hello @AH-Yussef, vava is correct. If the product is directly created (e.g. through an xml-rpc/another method creates it directly) or when there is a `'default_detailed_type': 'service'` used in a view for newly created products, then only the default 'receive' will be assigned and the compute won't be called unless the user changes the `detailed_type`. It is better to change the default to a lambda for consistent behavior in these cases.

**Pull request closed without merging**